sway-lib-std/src/experimental/storage.sw

```diff
@@ -0,0 +1,287 @@
+    ///
+    /// ```sway
+    /// fn foo() {
+    ///     let r: StorageHandle<u64> = StorageHandle {
+    ///         key: 0x0000000000000000000000000000000000000000000000000000000000000000,
+    ///         offset: 2,
+    ///     };
+    ///
+    ///     // Reads the third word from storage slot with key 0x000...0
+    ///     let x:Option<u64> = r.try_read();
+    /// }
+    /// ```
+    #[storage(read)]
+    pub fn try_read(self) -> Option<T> {
+        read(self.key, self.offset)
+    }
+
+    /// Writes a value of type `T` starting at the location specified by `self`. If the value
+    /// crosses the boundary of a storage slot, writing continues at the following slot.
+    ///
+    /// ### Arguments
+    ///
+    /// * value: the value of type `T` to write
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// fn foo() {
+    ///     let r: StorageHandle<u64> = StorageHandle {
+    ///         key: 0x0000000000000000000000000000000000000000000000000000000000000000,
+    ///         offset: 2,
+    ///     };
+    ///     let x = r.write(42); // Writes 42 at the third word of storage slot with key 0x000...0
+    /// }
+    /// ```
+    #[storage(read, write)]
+    pub fn write(self, value: T) {
+        write(self.key, self.offset, value);
+    }
+}
+
+/// A persistent key-value pair mapping struct.
+pub struct StorageMap<K, V> {}
+
+impl<K, V> StorageMap<K, V> {
+    /// Inserts a key-value pair into the map.
+    ///
+    /// ### Arguments
+    ///
+    /// * `key` - The key to which the value is paired.
+    /// * `value` - The value to be stored.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     map: StorageMap<u64, bool> = StorageMap {}
+    /// }
+    ///
+    /// fn foo() {
+    ///     let key = 5_u64;
+    ///     let value = true;
+    ///     storage.map.insert(key, value);
+    ///     let retrieved_value = storage.map.get(key);
+    ///     assert(value == retrieved_value);
+    /// }
+    /// ```
+    #[storage(read, write)]
+    pub fn insert(self: StorageHandle<Self>, key: K, value: V) {
+        let key = sha256((key, self.key));
+        write::<V>(key, 0, value);
+    }
+
+    /// Retrieves the `StorageHandle` that describes the raw location in storage of the value
+    /// stored at `key`, regardless of whether a value is actually stored at that location or not.
+    ///
+    /// ### Arguments
+    ///
+    /// * `key` - The key to which the value is paired.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     map: StorageMap<u64, bool> = StorageMap {}
+    /// }
+    ///
+    /// fn foo() {
+    ///     let key = 5_u64;
+    ///     let value = true;
+    ///     storage.map.insert(key, value);
+    ///     let retrieved_value = storage.map.get(key).read();
+    ///     assert(value == retrieved_value);
+    /// }
+    /// ```
+    #[storage(read)]
+    pub fn get(self: StorageHandle<Self>, key: K) -> StorageHandle<V> {
```

Comment:
Nice, I think returning the handle here makes a lot of sense and probably provides the most flexibility :+1: